week-03/assessment/solar-farm/solar-farm/src/main/java/learn/solar/ui/Controller.java

```diff
@@ -0,0 +1,103 @@
+package learn.solar.ui;
+
+import learn.solar.data.DataException;
+import learn.solar.domain.PanelResult;
+import learn.solar.domain.PanelService;
+import learn.solar.models.Panel;
+
+import java.util.List;
+
+//UI CONTROLLER
+public class Controller {
+    private View view;
+    private final PanelService service;
+
+
+    public Controller(View view, PanelService service) {
+        this.view = view;
+        this.service = service;
+    }
+    public void run() {
+        view.printHeader("Welcome to the Solar Farm!");
+        try {
+            runMenu();
+        }catch (DataException ex){
+            view.displayText("Error, something went wrong.");
+            view.displayText(ex.getMessage());
+        }
+        view.displayText("Goodbye!");
+    }
+    private void runMenu() throws DataException {
+        boolean exit = false;
+        while (!exit) {
+            int selection = view.chooseMenuOption();
+            switch (selection) {
+                case 1:
+                    addPanel();
+                    break;
+                case 2:
+                    updateEntry();
+                    break;
+                case 3:
+                    deletePanel();
+                    break;
+                case 4:
+                    viewBySection();
+                    break;
+                case 5:
+                    exit = true;
+                    break;
+            }
+        }
+    }
+        //- `private void addPanel()` -- coordinates between service and view to add a new panel
+        private void addPanel() throws DataException {
+            Panel panel = view.makePanel();
+            PanelResult result = service.add(panel);
+            if (result.isSuccess()) {
+                view.displayText("Panel successfully created!");
+                view.printResult(result);
+            } else {
+                view.displayErrors(result.getMessages());
+            }
+        }
+        //- `private void viewBySection()` -- coordinates between service and view to display panels in a section
+        private void viewBySection() throws DataException {
+            String section = view.readSection();
+            List<Panel> panels = service.findBySection(section);
+            view.printPanels(section, panels);
+        }
+        //- `private void updatePanel()` -- coordinates between service and view to update a panel
+        private void updateEntry() throws DataException {
+            String section = view.readSection();
+            List<Panel> panels = service.findBySection(section);
+            view.printPanels(section, panels);
+            Panel panel = view.choosePanel(section, panels);
+            view.update(panel);
+            if (panel == null ) {
+                System.out.println("aborting");
```

Comment:
system.out.println should never be used in the controller - consider a better error message here - we should be getting errors from our array of error messages in the result class